pymatgen/command_line/mcsqs_caller.py

```diff
@@ -29,93 +34,222 @@ def run_mcsqs(
+            p.kill()
+            p.communicate()
 
-        # Run mcsqs to find sqs structure
-        p = subprocess.Popen(
-            mcsqs_find_sqs_cmd
-        )
+        # Find the best sqs structures
+        if instances > 1:
 
-        try:
-            p.communicate(timeout=search_time * 60)
-            raise Exception("mcsqs exited before timeout reached")
-        except subprocess.TimeoutExpired:
-            p.kill()
+            if not os.path.exists("bestcorr1.out"):
+                raise RuntimeError("mcsqs did not generate output files, "
+                                   "is search_time sufficient or are number of instances too high?")
+
+            p = Popen(
+                ["mcsqs", "-best"]
+            )
             p.communicate()
-            if os.path.exists("bestsqs.out") and os.path.exists("bestcorr.out"):
-
-                # Convert output sqs structure to cif file
-                p = subprocess.Popen(
-                    "str2cif < bestsqs.out > bestsqs.cif",
-                    shell=True
-                )
-                p.communicate()
-
-                # Get objective function
-                with open('bestcorr.out', 'r') as f:
-                    lines = f.readlines()
-                objective_function = float(lines[-1].split('=')[-1].strip())
-
-                return Sqs(
-                    bestsqs=Structure.from_file("bestsqs.cif"),
-                    objective_function=objective_function
-                )
-
-            else:
-                raise TimeoutError("Cluster expansion took too long.")
+
+        if os.path.exists("bestsqs.out") and os.path.exists("bestcorr.out"):
+            sqs = _parse_sqs_path(".")
+            return sqs
+
+        else:
+            os.chdir(original_directory)
+            raise TimeoutError("Cluster expansion took too long.")
+
+
+def _parse_sqs_path(path) -> Sqs:
+    """
+    Private function to parse mcsqs output directory
+    Args:
+        path: directory to perform parsing
+
+    Returns:
+        Tuple of Pymatgen structure SQS of the input structure, the mcsqs objective function,
+            list of all SQS structures, and the directory where calculations are run
+    """
+
+    path = Path(path)
+
+    # instances will be 0 if mcsqs was run in parallel, or number of instances
+    instances = len(list(path.glob('bestsqs*[0-9]*')))
+
+    # Convert best SQS structure to cif file and pymatgen Structure
+    p = Popen(
+        "str2cif < bestsqs.out > bestsqs.cif",
+        shell=True,
+        cwd=path
+    )
+    p.communicate()
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        bestsqs = Structure.from_file(path / "bestsqs.cif")
+
+    # Get best SQS objective function
+    with open(path / 'bestcorr.out', 'r') as f:
+        lines = f.readlines()
+    # objective_function = float(lines[-1].split('=')[-1].strip())
+    # TODO: account for "Perfect_match" objective function
```

Comment:
I think this TODO was done(?) and the commented-out line above can be removed

Reply:
Done